lightning/src/util/persist.rs

```diff
@@ -0,0 +1,81 @@
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Traits for handling persistence
+
+use core::ops::Deref;
+use bitcoin::hashes::hex::ToHex;
+use prelude::*;
+use io::{self};
+
+use crate::{chain::{keysinterface::{Sign, KeysInterface}, self, transaction::{OutPoint}, chaininterface::{BroadcasterInterface, FeeEstimator}, chainmonitor::{Persist, MonitorUpdateId}, channelmonitor::{ChannelMonitor, ChannelMonitorUpdate}}, ln::channelmanager::ChannelManager, routing::network_graph::NetworkGraph};
+
+use super::{logger::Logger, ser::Writeable};
+
+/// Trait that handles persisting a [`ChannelManager`] and [`NetworkGraph`] to disk.
+pub trait Persister<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> where
+M::Target: 'static + chain::Watch<Signer>,
+T::Target: 'static + BroadcasterInterface,
+K::Target: 'static + KeysInterface<Signer = Signer>,
+F::Target: 'static + FeeEstimator,
+L::Target: 'static + Logger
+{
+	/// Persist the given ['ChannelManager'] to disk, returning an error if persistence failed.
+	fn persist_manager(&self, channel_manager: &ChannelManager<Signer, M, T, K, F, L>) -> Result<(), io::Error>;
+
+	/// Persist the given [`NetworkGraph`] to disk, returning an error if persistence failed.
+	fn persist_graph(&self, network_graph: &NetworkGraph) -> Result<(), io::Error>;
+}
+
+/// Trait for a key-value store for persisting some writeable object at some key
+pub trait KVStorePersister {
+	/// Persist the given writeable using the provided key
+	fn persist<W: Writeable>(&self, key: String, object: &W) -> io::Result<()>;
+	/// Get the key for persisting a channel manager
+	fn get_channel_manager_key(&self) -> String { String::from("manager") }
```

Comment:
Hmmmm, do users have a good reason to override this? ISTM its less work to just implement `Persister` if they want to bother with overriding the key. We might as well just remove the key-fetch functions and make them a part of the trait definition.

Reply:
It's not entirely clear if there's a good reason to override yet but it might be useful to know what the keys are when implementing KVStorePersister `persist`.  For example you might want to ignore errors when persisting the network_graph and just print a warning to console.  To do that you'd need to be able to know the key being used by persist_graph inside of `persist`.  

Without this you'd have to just hardcode the string in your `persist` implementation.  

When you say "make them part of the train definition" do you mean just hardcode the strings into the implementation of Persister/Persist?

Reply:
removed these for now